Add/407 merchant descriptor

vbelolapotkov opened the pull request:
Fixes #407 

**Note:** Server part (304-gh-woocommerce-payments-server) was already shipped.

#### Changes proposed in this Pull Request

* Allows merchant to edit statement descriptor on the plugin settings page

#### Testing instructions

**Happy path:**
* Launch server from 304-gh-woocommerce-payments-server
* Go to payment settings page. Current payment descriptor should be displayed matching one from Stripe Dashboard.
* Modify statement descriptor and save changes.
* Check changes applied on the site and in the Stripe Dashboard.
* Try to make a purchase. New statement descriptor should be applied.

**Edge cases:**
_Failed to fetch account info:_
* Throw exception either on the server (and clear cache) or in `get_cached_account_data`.
* Go to settings page. Field should not be rendered and exception message should be logged.

_Use invalid descriptor value:_
* Enter invalid descriptor value. See [docs](https://stripe.com/docs/statement-descriptors#requirements) for reference.
* Save changes.
* The dismissable error notice should be displayed at the top.

#### Screenshots
**Default view:**

<img width="1013" alt="Screenshot 2020-08-18 at 18 58 10" src="https://user-images.githubusercontent.com/3139099/90536592-ec9f0a80-e184-11ea-872b-7daa7aa203c7.png">

**With our own error message:** This message is message generated by our own validation logic and can be translated.

<img width="1603" alt="Screenshot 2020-08-18 at 19 00 11" src="https://user-images.githubusercontent.com/3139099/90536689-122c1400-e185-11ea-9f2a-b999657ea858.png">

**With Stripe's message:** This message is from Stripe API. Under normal conditions it won't be displayed to the user, but in case Stripe's validation rules change and invalid value slips through our validation similar message will be displayed.

<img width="1597" alt="image" src="https://user-images.githubusercontent.com/3139099/90537118-9aaab480-e185-11ea-8084-c7ca94d3a8cf.png">



-------------------

- [ ] Tested on mobile (or does not apply).
- [x] Tests are created (or not required).
- [x] Changelog entry is created.

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

thenbrent commented:
I've added the _Needs docs_ label to this one. We'll want to publish a public doc with guidance on good statement descriptors within the constraints, like [Stripe's doc](https://stripe.com/docs/statement-descriptors). We should also link to that doc from this setting. That doesn't need to be a blocker for this to feature ship in its initial form though.

DanReyLop commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -618,6 +624,60 @@ public function generate_account_status_html() {
+		}
+
+		// Value from option is used as input value. Update option to make sure proper value is displayed.
+		$this->update_option( $key, $descriptor );
```
This ends up writing the descriptor to the `woocommerce_woocommerce_payments_settings` option in the `wp_options` table. The same will happen when calling `parent::process_admin_options()`. That means that the descriptor will be both stored in the database (but never read) and fetched from the WCPay server. That could cause problems.

There are some ways to prevent it. You could filter the result of the `get_option` call [here](https://github.com/woocommerce/woocommerce/blob/14a26aca2cde9c81ade02bdc6dd0a34e184a4202/includes/abstracts/abstract-wc-settings-api.php#L262) to inject the descriptor value (there's a `option_OPTION_NAME` filter called at the end of the `get_option` function). There's also [a filter](https://github.com/woocommerce/woocommerce/blob/14a26aca2cde9c81ade02bdc6dd0a34e184a4202/includes/abstracts/abstract-wc-settings-api.php#L219) to "cleanup" any settings before there are written into the database.

What do you think about those changes? They're not a deal-breaker because the descriptor is *never* read from the database, so don't feel like you need to do them.

There's also a small issue with your approach. The descriptor stored in the database would be returned if the WCPay settings are fetched through a REST API endpoint (the `generate_*_html` method isn't called there).

vbelolapotkov replied:
> You could filter the result of the get_option call here to inject the descriptor value (there's a option_OPTION_NAME filter called at the end of the get_option function)

This approach didn't work because `get_option` called at the time when settings are initialized and using account instance to fetch account data throws because GW initializations is not finished yet (circular dependency problem).
As an alternative solution, `get_option` method was overrided to get value from account instead of plugin options.

> There's also a filter to "cleanup" any settings before there are written into the database.

Nice suggestion, used it to handle option update in connected account.

Fix is in https://github.com/Automattic/woocommerce-payments/pull/788/commits/d3650576650c45fe50a4e68bd08b32da7901ecc6

DanReyLop commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -618,6 +624,60 @@ public function generate_account_status_html() {
+	public function validate_account_statement_descriptor_field( $key, $value ) {
+		$sanitized_value = $this->validate_text_field( $key, $value );
+
+		// Validation can be done with a single regex but splitting into multiple for better readability.
```
Awesome!

DanReyLop commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -87,35 +87,40 @@ public function __construct(
 			],
-			'manual_capture'  => [
+			'account_statement_descriptor' => [
+				'type'  => 'account_statement_descriptor',
+				'title' => __( 'Statement Descriptor', 'woocommerce-payments' ),
+				'label' => __( 'Statement descriptor to display in receipt', 'woocommerce-payments' ),
```
`label` is not rendered on text fields.
```suggestion
				'type'        => 'account_statement_descriptor',
				'title'       => __( 'Statement Descriptor', 'woocommerce-payments' ),
				'description' => __( 'Statement descriptor to display in receipt.', 'woocommerce-payments' ),
```

vbelolapotkov replied:
👍 Fixed in https://github.com/Automattic/woocommerce-payments/pull/788/commits/9341cf604063d5a19c895cc66b3c13cc36740146

vbelolapotkov commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -717,6 +724,129 @@ public function generate_account_status_html() {
+	public function sanitize_plugin_settings( $settings ) {
+		if ( isset( $settings['account_statement_descriptor'] ) ) {
+			$this->update_statement_descriptor( $settings['account_statement_descriptor'] );
+			$settings['account_statement_descriptor'] = '';
```
Don't think so, it's prefilled in `init_settings` with `''` [here](https://github.com/woocommerce/woocommerce/blob/14a26aca2cde9c81ade02bdc6dd0a34e184a4202/includes/abstracts/abstract-wc-settings-api.php#L267).

vbelolapotkov commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -717,6 +724,129 @@ public function generate_account_status_html() {
+			! preg_match( $no_specials, $sanitized_value )
+		) {
+			throw new Exception(
+				__( 'Invalid Statement descriptor. Statement descriptor should be between 5 and 22 characters long, contain at least single Latin character and does not contain special characters: \' " *', 'woocommerce-payments' )
```
Yes, I've checked the docs and experimented myself that's where real rules come from I decided to allow anything Stripe accepts 🙂 I don't see any reason why we would add extra restrictions.

LevinMedia commented:
@vbelolapotkov awesome, thank you! 

Can you update the label on the left from "Statement descriptor" to: 
"Customer Bank Statement" 

Then update the helper text under the field to read: 
"Edit the way your store name appears on your customers’ bank statements."

This next bit is probably out of scope for this specific issue - but I think we should consider opening another issue and making this two fields - one for the description, and a second for the store's phone number, which would be combined into one line on the billing statement. 

Ideally we'd automatically fill both fields from previously supplied data, the website's URL, or the business name, and their phone number. Including the phone number would give customers the option to call the merchant (hopefully) before submitting a chargeback if they still didn't recognize the descriptor. cc @thenbrent

vbelolapotkov commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -717,6 +724,129 @@ public function generate_account_status_html() {
+	 * @throws Exception When statement descriptor is invalid.
+	 */
+	public function validate_account_statement_descriptor_field( $key, $value ) {
+		$sanitized_value = $this->validate_text_field( $key, $value );
```
Good point! Fixed in https://github.com/Automattic/woocommerce-payments/pull/788/commits/38a5770b28350528e15c35dfbc4c94365bedcec0.

vbelolapotkov commented:
> Can you update the label on the left from "Statement descriptor" to:
> "Customer Bank Statement"

Changed to `Customer bank statement` to comply with uppercasing only first word letter. And updated error message accordingly: `Customer bank statement is invalid. Statement should be between 5 and 22 characters long, contain at least single Latin character and does not contain special characters: ' " * < >`.

> Then update the helper text under the field to read:
> "Edit the way your store name appears on your customers’ bank statements."

Done.

> This next bit is probably out of scope for this specific issue - but I think we should consider opening another issue and making this two fields - one for the description, and a second for the store's phone number, which would be combined into one line on the billing statement.

Are you sure about that? Statement descriptor is limited by 22 chars only [[docs](https://stripe.com/docs/statement-descriptors#requirements)]. Phone number will eat at least half of it leaving almost zero space for the business name.
We might think about moving it to another issue as well, because if we introduce phone number later and combine both to make a statement descriptor we might end up above that 22 chars limit.

> Ideally we'd automatically fill both fields from previously supplied data, the website's URL, or the business name, and their phone number. 

This setting value is taken directly from Stripe account. It's not stored anywhere on the site (only in cached account data). Not sure though how it will play if we want to combine it with business phone number stored in Stripe account.

LevinMedia commented:
@vbelolapotkov I wasn't aware of the character limit. Thanks for pointing that out. Good to go from my end for now.

thenbrent commented:
> Ideally we'd automatically fill both fields from previously supplied data, the website's URL, or the business name, and their phone number.

Stripe do that already for us.

> >     we should consider opening another issue and making this two fields - one for the description, and a second for the store's phone number, which would be combined into one line on the billing statement.
> Are you sure about that? Statement descriptor is limited by 22 chars only

Depending on the bank, both may be included on the statement but the phone number doesn't need to be part of the statement descriptor. If the bank supports it, the phone number will be added to the statement in addition to whatever the value of the statement descriptor is.

vbelolapotkov commented:
@marcinbot I think it's ready for the another round of review.

vbelolapotkov commented:
@thenbrent I've created another task for the docs and added it to the current sprint scope.
